api/v1alpha1/osbuildconfig_types.go

```diff
@@ -0,0 +1,133 @@
+
+// OSBuildConfigSpec defines the desired state of OSBuildConfig
+type OSBuildConfigSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Details defines what to build
+	Details BuildDetails `json:"build_details"`
+	// Triggers defines when to build
+	Triggers BuildTriggers `json:"build_triggers"`
+}
+
+// BuildDetails includes all the information needed to build the image
+type BuildDetails struct {
+	// Distribution is the name of the O/S distribution
+	Distribution string `json:"distribution"`
+	// Customizations defines the changes to be applied on top of the base image (optional)
+	Customizations *Customizations `json:"customizations,omitempty"`
+	// TargetImage defines the requested output image
+	TargetImage TargetImage `json:"target_image"`
+}
+
+// Customizations defines the changes to be applied on top of the base image
+type Customizations struct {
+	// Packages is a list of RPM packages to install (optional)
+	Packages []string `json:"packages,omitempty"`
+	// Users is the list of Users to add to the image (optional)
+	Users []User `json:"users,omitempty"`
+	// Services defines the services to enable or disable (optional)
+	Services *Services `json:"services,omitempty"`
+}
+
+// User defines a single user to be configured
+type User struct {
+	// Name is the username for the new user
+	Name string `json:"name"`
+	// Groups is the groups to add the user to (optional)
+	Groups []string `json:"groups,omitempty"`
+	// Key is the user's SSH public key (optional)
+	Key *string `json:"key,omitempty"`
+}
+
+type Services struct {
+	// Enabled is the list of services to enabled (optional)
+	Enabled []string `json:"enabled,omitempty"`
+	// Disabled is the list of services to disabled (optional)
+	Disabled []string `json:"disabled,omitempty"`
+}
+
+type TargetImage struct {
+	// Architecture defines target architecture of the image
+	Architecture string `json:"architecture"`
+	// ImageType defines the target image type
+	TargetImageType TargetImageType `json:"target_image_type"`
+	// OSTree is the OSTree configuration of the build (optional)
+	OSTree *OSTreeConfig `json:"ostree,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=edge-container;edge-installer
+type TargetImageType string
+
+// OSTreeConfig defines the OSTree ref details
+type OSTreeConfig struct {
+	// Url is the URL of the target build
+	Url *string `json:"url"`
+	// Ref is the ref of the target build
+	Ref *string `json:"ref"`
+	// Parent is the ref of the parent of target build (Optional)
+	Parent *string `json:"parent"`
+}
+
+type BuildTriggers struct {
+	// ConfigChange if True trigger a new build upon any change in this BuildConfig CR (optional)
+	ConfigChange *bool `json:"config_change,omitempty"`
+	// WebHook defines the way to trigger a build using a REST call (optional)
```

Comment:
The idea was to mimic the generic webhook trigger os BuildConfig as described here: https://docs.openshift.com/container-platform/4.7/cicd/builds/triggering-builds-build-hooks.html#builds-using-generic-webhooks_triggering-builds-build-hooks.
But, since we cannot (or maybe just shouldn't) tap directly into the OpenShift API, as you said, it will require us to expose a service.
Now, the reason this option is needed is that since the packages list does not state the version, a user will want to trigger a build without changing the CR itself. This new build will generate a new image because of newer pacakges.

Reply:
In addition to not forcing the user to create new CRs for minor changes, this design provides two things:
1. Build History - the user can go over the OSBuild CRs created from an OSBuildConfig and see the changes
2. OSTree parent - as of the second build, unless instructed otherwise, the build will set the previous ref as its parent

Reply:
Because if we plan to mimic the behaviour of `BuildConfig`, then a build should be either of type *-commit/container or *-installer, right? Sorry if I'm missing something here...

Reply:
In the current design an OSBuildConfig produces one type of image. So, like you wrote in the last comment, it's one of `commit`, `container` or `installer` (out of these three options). Please note that for now, I've only set `container` and `installer` as these are the use cases that we are focusing on

Reply:
After discussing the @jordigilh he explained that in order to generate an ISO file (type installer), the commit has to be serviced somewhere. So, in order to generate an installer, the operator will first have to generate a commit (using the container), service it and generate the ISO. This means that if the user requests an ISO, she will get both the commit container and the ISO.